packages/core/src/nodes.rs

```diff
@@ -1143,6 +1083,137 @@ pub trait HasAttributes {
     ) -> Self;
 }
 
+#[derive(Clone)]
+pub struct NodeCursor<'a> {
+    // The position in the inner node, represented as a list of child offsets
+    position: Vec<u8>,
+    // The inner node we are moving in. This is a block of static nodes with references to any dynamic children
+    inner: &'a VNode,
+    // An optional reference to the virtual dom, for accessing children that are rendered
+    vdom: Option<&'a VirtualDom>,
+}
+
+impl<'a> NodeCursor<'a> {
+    /// Creates a new NodeCursor on the root node
+    pub fn new(inner: &'a VNode) -> NodeCursor<'a> {
+        NodeCursor { position: vec![0], vdom: None, inner }
+    }
+
+    /// Allows the NodeCursor to access any dynamic nodes through the VirtualDom
+    pub fn with_virtualdom(mut self: NodeCursor<'a>, vdom: &'a VirtualDom) -> NodeCursor<'a> {
+        self.vdom = Some(vdom);
+        self
+    }
+
+    /// The current node the cursor is on
+    pub fn current_node(&self) -> TemplateNode {
+        self.try_current_node().unwrap()
+    }
+
+    fn try_current_node(&self) -> Option<TemplateNode> {
+        let mut child_index_iter = self.position.iter().copied();
+        let mut current =
+            self.inner.template.get().roots.get(child_index_iter.next().unwrap() as usize)?;
+
+        for child_index in child_index_iter {
+            match current {
+                TemplateNode::Element { children, .. } => {
+                    current = children.get(child_index as usize)?
+                }
+                TemplateNode::Dynamic { id } => match &self.inner.dynamic_nodes[*id] {
+                    DynamicNode::Fragment(children) => {
+                        current = &children.get(child_index as usize)?.template.get().roots[0]
```

Comment:
Dynamic nodes are always a leaf in the template subtree tree. If we reach a fragment, we need to traverse into the child VNode